### PR TITLE
Seperate KGB-server and KGB rules to allow git commit hook creation

### PR DESCRIPTION
**Pull request by aborsu**

### CODE REVIEW

**Comment by SagaStalin** [comment]
<KGB>cf5cfeb6d7741d09226f2d2d602afbf484ecb4c0
>fix(linter): Lint code(flake8)

Conform status to `<type>(<scope>): <subject>` pattern
type and scope cannot contain spaces or parenthesisMore info [here](https://github.com/sagacify/komitet-gita-bezopasnosti)

**Comment by SagaStalin** [comment]
<KGB>4794ff5f80d1701dd5b85bc58e7a0abbf2acfd26
>chore(cheeseshop): Package

Status must start with one of the following types:
*doc, feat, fix, perf, refactor, revert, style, test, version*More info [here](https://github.com/sagacify/komitet-gita-bezopasnosti)

File: kgbrules/config.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from os import environ as env
```
Comment (SagaStalin) [inline kgbrules/config.py]: MAJOR Add a docstring to this module. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1720)
MINOR Missing module docstring [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0111)

File: kgbrules/raw_rules.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (SagaStalin) [inline kgbrules/raw_rules.py]: MINOR Missing module docstring [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0111)

File: kgbrules/raw_rules.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+def check_for_tabs(commit_message):
```
Comment (SagaStalin) [inline kgbrules/raw_rules.py]: MAJOR Add a docstring to this module. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1720)

File: kgbrules/line_rules.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+from .config import MAX_LINE_LENGTH
```
Comment (SagaStalin) [inline kgbrules/line_rules.py]: MAJOR Add a docstring to this module. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1720)
MINOR Missing module docstring [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0111)

File: kgbrules/status_rules.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
+import re
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MAJOR Add a docstring to this module. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1720)
MINOR Missing module docstring [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0111)

File: kgbrules/status_rules.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    >>> print(check_status_formatting("fix(this):n  "))
+    Add space after colon.
+    Uppercase the first character of the subject.
+    Make subject at least three character long.
+    Strip trailing whitespaces from status line.
+    """
+    if _is_merge(status_line):
+        return None
+    match = status_pattern.match(status_line)
+    if (match is None):
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MAJOR Remove those useless parentheses [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1110)
MAJOR Remove the parentheses after this "if" keyword [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1721)
MINOR Unnecessary parens after 'if' keyword [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0325)

File: kgbrules/status_rules.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
+import re
+
+from .config import MAX_STATUS_LENGTH
+from .config import TYPES
+
+status_pattern = re.compile(
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MINOR Invalid constant name "status_pattern" [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0103)

File: kgbrules/status_rules.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    """^([^()\s]*)(\s*)\(([^()\s]*)\)(\s*):(\s*)(.*\w)(\s*)$""")
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MINOR Anomalous backslash in string: '\s'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\s'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\('. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\s'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\)'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\s'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\s'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\w'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\s'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)

File: kgbrules/status_rules.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+spaces = re.compile("\w*\s+\w*\(|\w+\(.*[^\w].*\):")
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MAJOR Remove this "\", add another "\" to escape it, or make this a raw string. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1717)
MINOR Anomalous backslash in string: '\w'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\s'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\w'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\('. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\w'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\('. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\w'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Anomalous backslash in string: '\)'. String constant might be missing an r prefix. [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW1401)
MINOR Invalid constant name "spaces" [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0103)

File: kgbrules/status_rules.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+def check_status_length(status_line):
+    """Check status length.
+
+    >>> check_status_length("this is short enough")
+
+    >>> check_status_length("TOO_LONG" * 10)
+    'Limit status to 50 characters'
+    """
+    if not _is_merge(status_line):
+        if len(status_line) > MAX_STATUS_LENGTH:
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1066)

File: kgbrules/status_rules.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+from .config import MAX_STATUS_LENGTH
+from .config import TYPES
+
+status_pattern = re.compile(
+    """^([^()\s]*)(\s*)\(([^()\s]*)\)(\s*):(\s*)(.*\w)(\s*)$""")
+
+spaces = re.compile("\w*\s+\w*\(|\w+\(.*[^\w].*\):")
+
+_type_string = ", ".join(sorted(TYPES))
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MINOR Invalid constant name "_type_string" [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0103)

File: kgbrules/status_rules.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def _is_merge(status_line):
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MAJOR Add a docstring to this function. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1720)

File: kgbrules/status_rules.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+    >>> check_status_length("TOO_LONG" * 10)
+    'Limit status to 50 characters'
+    """
+    if not _is_merge(status_line):
+        if len(status_line) > MAX_STATUS_LENGTH:
+            return "Limit status to {0} characters".format(MAX_STATUS_LENGTH)
+
+
+def check_status_formatting(status_line):
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MAJOR Function has a complexity of 15 which is greater than 10 authorized. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AFunctionComplexity)
MAJOR This function has 4 returns or yields, which is more than the 3 allowed. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1142)

File: kgbrules/__init__.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+from . import raw_rules
+from . import status_rules
+
+
+def _get_rules(module):
+    """Get all functions from a module."""
+    return [tup[1] for tup in getmembers(module, isfunction)
+            if tup[0][0] != "_"]
+
+_line_rules = _get_rules(line_rules)
```
Comment (SagaStalin) [inline kgbrules/__init__.py]: MINOR Invalid constant name "_line_rules" [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0103)

File: kgbrules/__init__.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+_raw_rules = _get_rules(raw_rules)
```
Comment (SagaStalin) [inline kgbrules/__init__.py]: MINOR Invalid constant name "_raw_rules" [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0103)

File: kgbrules/__init__.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+_status_rules = _get_rules(status_rules)
```
Comment (SagaStalin) [inline kgbrules/__init__.py]: MINOR Invalid constant name "_status_rules" [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0103)

File: kgbrules/__init__.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def split_lines(commit_message):
```
Comment (SagaStalin) [inline kgbrules/__init__.py]: MAJOR Add a docstring to this function. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1720)
MINOR Missing function docstring [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0111)

File: kgbrules/__init__.py [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+    commit_lines = split_lines(commit_message)
+    for rule in _status_rules:
+        err = rule(commit_lines[0])
+        if err is not None:
+            errors.append(err)
+    for check in _line_rules:
+        err = check(commit_lines)
+        if err is not None:
+            errors.append(err)
+    return(errors)
```
Comment (SagaStalin) [inline kgbrules/__init__.py]: MAJOR Remove the parentheses after this "return" keyword [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1721)
MAJOR Remove those useless parentheses [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1110)
MINOR Unnecessary parens after 'return' keyword [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AC0325)

File: kgbrules/__init__.py [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+_line_rules = _get_rules(line_rules)
+_raw_rules = _get_rules(raw_rules)
+_status_rules = _get_rules(status_rules)
+
+
+def split_lines(commit_message):
+    return re.split('\r?\n', commit_message)
+
+
+def apply(commit_message):
```
Comment (SagaStalin) [inline kgbrules/__init__.py]: MINOR Redefining built-in 'apply' [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW0622)

**Comment by SagaStalin** [comment]
SonarQube analysis reported 61 issues
* MAJOR 30 major
* MINOR 31 minor

Watch the comments in this conversation to review them.

**Comment by SagaStalin** [comment]
<KGB>4794ff5f80d1701dd5b85bc58e7a0abbf2acfd26
>chore(cheeseshop): Package

Status must start with one of the following types:
*doc, feat, fix, perf, refactor, revert, style, test, version*More info [here](https://github.com/sagacify/komitet-gita-bezopasnosti)

File: kgbrules/status_rules.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+STATUS_PATTERN = re.compile(
+    r"^([^()\s]*)(\s*)\(([^()\s]*)\)(\s*):(\s*)(.*\w)(\s*)$")
+
+_TYPE_STRING = ", ".join(sorted(TYPES))
+
+
+def _is_merge(status_line):
+    return status_line.startswith("Merge pull request #")
+
+def _matches_pattern(status_line)
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MAJOR Parse error at line 17 column 33:

   11: join(sorted(TYPES))
   12: 
   13: 
   14: def _is_merge(status_line):
   15:     return status_line.startswith("Merge pull request #")
   16: 
  -->  def _matches_pattern(status_line)
   18:     match = STATUS_PATTERN.match(status_line)
   19:     if match is None:
   20:         return ("Conform status to `<type>(<scope>): <subject>` pattern\n" +
   21:                 "type and scope cannot contain spaces or parenthesis")
   22:     errors = [] [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AParsingError)
MAJOR invalid syntax [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AE0001)

File: kgbrules/__init__.py [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+_RAW_RULES = _get_rules(raw_rules)
+_STATUS_RULES = _get_rules(status_rules)
+
+
+def split_lines(commit_message):
+    """Splits lines of the commit"""
+    return re.split('\r?\n', commit_message)
+
+
+def apply(commit_message):
```
Comment (SagaStalin) [inline kgbrules/__init__.py]: MINOR Redefining built-in 'apply' [rule](https://sonar.sagacify.com/coding_rules#rule_key=Pylint%3AW0622)

**Comment by SagaStalin** [comment]
SonarQube analysis reported 3 issues
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SagaStalin** [comment]
<KGB>4794ff5f80d1701dd5b85bc58e7a0abbf2acfd26
>chore(cheeseshop): Package

Status must start with one of the following types:
*doc, feat, fix, perf, refactor, revert, style, test, version*More info [here](https://github.com/sagacify/komitet-gita-bezopasnosti)

File: kgbrules/status_rules.py [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+    r"^([^()\s]*)(\s*)\(([^()\s]*)\)(\s*):(\s*)(.*\w)(\s*)$")
+
+_TYPE_STRING = ", ".join(sorted(TYPES))
+
+
+def _is_merge(status_line):
+    return status_line.startswith("Merge pull request #")
+
+
+def _matches_pattern(status_line):
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MAJOR Function has a complexity of 13 which is greater than 10 authorized. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AFunctionComplexity)
MAJOR Add a docstring to this function. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AS1720)

File: kgbrules/status_rules.py [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+    >>> check_status_length("this is short enough")
+
+    >>> check_status_length("TOO_LONG" * 10)
+    'Limit status to 50 characters'
+    """
+    if len(status_line) > MAX_STATUS_LENGTH:
+        return "Limit status to {0} characters".format(MAX_STATUS_LENGTH)
+
+
+def check_status_formatting(status_line):
```
Comment (SagaStalin) [inline kgbrules/status_rules.py]: MAJOR Function has a complexity of 13 which is greater than 10 authorized. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AFunctionComplexity)

**Comment by ghost** [comment]
SonarQube analysis reported 1 issue
- MAJOR 1 major

Watch the comments in this conversation to review them.

File: kgbrules/status_rules.py (hunk 23 above)
Comment (aborsu) [reply kgbrules/status_rules.py]: After consideration. 
It doesn't seem to make sense to split this function into simpler components.
I believe it is understandable and will leave it as is.

File: kgb_rules/status_rules.py [hunk 24]
```diff
@@ -23,9 +19,8 @@ def check_status_length(status_line):
...
     'Limit status to 50 characters'
     """
-    if not _is_merge(status_line):
-        if len(status_line) > MAX_STATUS_LENGTH:
-            return "Limit status to {0} characters".format(MAX_STATUS_LENGTH)
+    if len(status_line) > MAX_STATUS_LENGTH:
+        return "Limit status to {0} characters".format(MAX_STATUS_LENGTH)
 
 
 def check_status_formatting(status_line):
```
Comment (ghost) [inline kgb_rules/status_rules.py:26]: MAJOR Function has a complexity of 13 which is greater than 10 authorized. [rule](https://sonar.sagacify.com/coding_rules#rule_key=python%3AFunctionComplexity)

**Pull request merged**